File: python/data_cleaner.py
"""Utilities for normalising glider data to OG1.0-compatible values."""

import re
from typing import Any

import pandas as pd


ESSENTIAL_COLUMNS = ("latitude", "longitude", "time")
ALIASES = {
    "latitude": ("latitude", "lat", "lat_deg", "latitude_deg"),
    "longitude": ("longitude", "lon", "lon_deg", "longitude_deg"),
    "time": ("time", "iso_time", "timestamp"),
}


def _snake_case(name: Any) -> str:
    """Return a strict snake_case field name."""
    value = re.sub(r"([a-z0-9])([A-Z])", r"\1_\2", str(name))
    value = re.sub(r"[^A-Za-z0-9]+", "_", value).strip("_").lower()
    return value or "unnamed"


def _validation_report(data: pd.DataFrame) -> dict[str, Any]:
    """Check the essential OG1.0 fields and return a serialisable report."""
    errors = [
        f"missing essential column: {column}"
        for column in ESSENTIAL_COLUMNS
        if not any(alias in data.columns for alias in ALIASES[column])
    ]
    warnings = []
    for column in ESSENTIAL_COLUMNS:
        present = next((alias for alias in ALIASES[column] if alias in data.columns), None)
        if present and data[present].isna().any():
            warnings.append(f"{column} contains missing or malformed values")
    return {
        "valid": not errors and not warnings,
        "errors": errors,
        "warnings": warnings,
    }


def clean_glider_data(df: pd.DataFrame) -> pd.DataFrame:
    """Normalise column names, timestamps, coordinates, and schema metadata.

    The cleaned frame is returned even when validation fails.  The explicit
    ``validation`` report is available at ``cleaned.attrs["validation"]``.
    """
    if not isinstance(df, pd.DataFrame):
        raise TypeError("df must be a pandas DataFrame")

    cleaned = df.copy()
    cleaned.columns = [_snake_case(column) for column in cleaned.columns]

    time_column = next((alias for alias in ALIASES["time"] if alias in cleaned.columns), None)
    if time_column:
        parsed_time = pd.to_datetime(cleaned[time_column], errors="coerce", utc=True)
        cleaned[time_column] = parsed_time.dt.strftime("%Y-%m-%dT%H:%M:%SZ")
        cleaned.loc[parsed_time.isna(), time_column] = pd.NA

    coordinate_columns = {
        coordinate: next(
            (alias for alias in aliases if alias in cleaned.columns), None
        )
        for coordinate, aliases in ALIASES.items()
        if coordinate in ("latitude", "longitude")
    }
    for coordinate, column in coordinate_columns.items():
        cleaned[column] = pd.to_numeric(cleaned[column], errors="coerce")

    if coordinate_columns.get("latitude"):
        column = coordinate_columns["latitude"]
        cleaned.loc[~cleaned[column].between(-90, 90), column] = pd.NA
    if coordinate_columns.get("longitude"):
        column = coordinate_columns["longitude"]
        cleaned.loc[~cleaned[column].between(-180, 180), column] = pd.NA

    cleaned.attrs["validation"] = _validation_report(cleaned)
    return cleaned

File: R/data_cleaner.R
# Utilities for normalising glider data to OG1.0-compatible values.

snake_case <- function(names) {
  names <- gsub("([a-z0-9])([A-Z])", "\\1_\\2", names, perl = TRUE)
  names <- gsub("[^A-Za-z0-9]+", "_", names)
  names <- tolower(gsub("^_|_$", "", names))
  names[names == ""] <- "unnamed"
  names
}

clean_glider_data <- function(df) {
  if (!is.data.frame(df)) {
    stop("df must be a data.frame")
  }

  names(df) <- snake_case(names(df))
  validation_errors <- character()
  validation_warnings <- character()

  aliases <- list(
    latitude = c("latitude", "lat", "lat_deg", "latitude_deg"),
    longitude = c("longitude", "lon", "lon_deg", "longitude_deg"),
    time = c("time", "iso_time", "timestamp")
  )
  essential_columns <- names(aliases)
  present_columns <- vapply(aliases, function(options) {
    any(options %in% names(df))
  }, logical(1))
  missing_columns <- essential_columns[!present_columns]
  if (length(missing_columns) > 0) {
    validation_errors <- paste("missing essential column:", missing_columns)
  }

  coordinate_columns <- vapply(aliases[c("latitude", "longitude")], function(options) {
    options[options %in% names(df)][1]
  }, character(1))
  for (column in coordinate_columns[!is.na(coordinate_columns)]) {
    values <- suppressWarnings(as.numeric(df[[column]]))
    values[!is.finite(values) | values < if (column == "latitude") -90 else -180 |
      values > if (column == "latitude") 90 else 180] <- NA_real_
    df[[column]] <- values
    if (anyNA(values)) {
      validation_warnings <- c(
        validation_warnings,
        paste(column, "contains missing or malformed values")
      )
    }
  }

  time_column <- aliases$time[aliases$time %in% names(df)][1]
  if (!is.na(time_column)) {
    parsed_time <- as.POSIXct(df[[time_column]], format = "%Y-%m-%dT%H:%M:%S", tz = "UTC")
    parsed_time[is.na(parsed_time)] <- as.POSIXct(
      df[[time_column]][is.na(parsed_time)], tz = "UTC"
    )
    df[[time_column]] <- ifelse(
      is.na(parsed_time),
      NA_character_,
      format(parsed_time, "%Y-%m-%dT%H:%M:%SZ", tz = "UTC")
    )
    if (anyNA(parsed_time)) {
      validation_warnings <- c(
        validation_warnings,
        "time contains missing or malformed values"
      )
    }
  }

  attr(df, "validation") <- list(
    valid = length(validation_errors) == 0 && length(validation_warnings) == 0,
    errors = validation_errors,
    warnings = validation_warnings
  )
  df
}
